### PR TITLE
RSDK-4716 - allow for graceful handling of failed modules

### PR DESCRIPTION
**Pull request by stuqdog**
#### Major Changes
Shorten the timeout for adding Modules from a default of 1m to 1s. When at 1m, a malformed module that would never start would effectively hang forever on a loop looking for the socket, causing `Reconfigure` calls to stall until we hit the 90 second `hungShutdownDeadline`, which kills all processes.

Additionally, we add a cancelFunc to the ctx that goes into the `startProcess` call, so that we can stop the loop immediately when closing the robot or the module manager.

### CODE REVIEW

**Comment by stuqdog** [comment]
@benjirewis @cheukt fyi I changed this to a draft PR because I believe there are issues to still be resolved before this can be merged. See slack message for details.

**Comment by benjirewis** [comment]
I'll pull this branch down and investigate a bit locally per offline discussion.

**Comment by stuqdog** [comment]
okay this is ready for review again. @benjirewis @cheukt to reiterate from offline conversation: with a malformed module, shutdown does still block on `Reconfigure` completing, which can take up to the resource building timeout (default 1m) per failed module. With enough failed modules this will instead hit the 90s web timeout, and everything will shut down at that time.

Additionally attempting to sigterm before the first `Reconfigure` call completes means we shut down the web server before it started, which logs an error. I think this is fine/expected.

**Comment by cheukt** [comment]
haven't looked deeply, but is there a way to add a test for this? definitely worried about future regressions

**Comment by stuqdog** [comment]
> haven't looked deeply, but is there a way to add a test for this? definitely worried about future regressions

yeah good question! I spent some time thinking on it and couldn't think of a good way to test. We're dealing with a number of pretty specific interactions (sigterm mid-Reconfigure, the precise timing of that sigterm relative to the 1m timeout and the 90s web timeout) that I couldn't think of/find good ways to mock. I can definitely keep looking though!

**Review by benjirewis — CHANGES_REQUESTED** [review]
Nice! Definitely like the direction this is headed in.

File: robot/impl/resource_manager.go
Comment (stuqdog) [reply robot/impl/resource_manager.go]: We want to continue updating the graph because we need resources to be in the graph so we can stop them. Without updating, I ran into a case where, when I sigtermed before the first Reconfigure was completed, a data manager wasn't put in the graph and so wasn't stopped in `robot.Close()` and so was leaking its `uploadData` goroutine. By still updating the graph, the resource manager becomes aware of the data manager (or any other resource with a goroutine) and can stop it appropriately.

Comment (benjirewis) [reply robot/impl/resource_manager.go]: Ah so sigterming a robot will no longer stop building goroutines?

Comment (stuqdog) [reply robot/impl/resource_manager.go]: I'm not aware that we've fundamentally changed how goroutines building works here, I think it's possible I ran into a race where the ctx was canceled in one thread and in another one we started up a goroutine for the data manager, but then because the ctx was canceled in the first thread the data manager was never added to the resource graph? The sigterm will definitely kill the data manager goroutine provided the resource manager knows about it, but that requires updating the graph.

File: robot/impl/local_robot.go
Comment (benjirewis) [reply robot/impl/local_robot.go:1067]: Gotcha. I realize you didn't write this code, but I don't see why we should ever be adding modules twice. If I remove the [section](https://github.com/viamrobotics/rdk/blob/main/robot/impl/resource_manager.go#L921-L931) of `updateResources` that adds modules from `conf.Added.Modules`, tests still seem to pass. `updateResources` is indeed only called by `Reconfigure`, and `Reconfigure` already adds modules to the resource manager... Can we just get rid of that section of `updateResources` and this `failedModules` parameter?

Comment (stuqdog) [reply robot/impl/local_robot.go]: yep, done!

File: module/modmanager/manager.go
Comment (cheukt) [inline module/modmanager/manager.go:147]: do we also need the lock here?

File: module/modmanager/manager.go
Comment (cheukt) [inline module/modmanager/manager.go:75]:
```suggestion
	// inStartup stores whether or not the manager or the OnUnexpectedExit function is trying
	// to startup this module; inStartupLock guards the execution of
	// an OnUnexpectedExit function for this module.
```
kind of?

**Review by cheukt — COMMENTED** [review]
so what's the current behavior? curious to see the scenarios that were tested and the outcomes, maybe in a table if that's not too difficult

**Comment by cheukt** [comment]
might be hard, but can we test that on manager closing, add/on unexpected exit handlers are properly exiting?

File: module/modmanager/manager.go
Comment (stuqdog) [reply module/modmanager/manager.go:147]: I think we want to leave the `inStartupLock` alone so that the OUE handler can take it and return quickly. Otherwise, we'll still end up with the OUE handler hanging on trying to get the lock until `add` finishes.

Comment (benjirewis) [reply module/modmanager/manager.go:147]: ^ yeah we don't want `add` to grab that lock here. If it did, then the OUE will succeed in getting the lock after `add` finishes, find that `inStartup` is now false, and attempt to restart the module that might have started successfully already.

File: module/modmanager/manager.go
Comment (benjirewis) [inline module/modmanager/manager.go:81]: I think this should still be called `inRecoveryLock`, it's only relevant for OUEs and stops them from overlapping with each other. Isn't used in `add`/during startup.

File: module/modmanager/manager.go
Comment (stuqdog) [reply module/modmanager/manager.go]: Pushed this change (wanted `manager of` not `manager or`)!

**Comment by stuqdog** [comment]
> so what's the current behavior? curious to see the scenarios that were tested and the outcomes, maybe in a table if that's not too difficult

I think right now there are basically two classes of cases when there's a malformed module:
  1. We attempt to sigterm in the middle of a `Reconfigure` call. In this case, we're blocked on either `Reconfigure` finishing or hitting the 90 second timeout, whichever comes first. Assuming one failed module and no modification to the default timeout, this means it can take up to a minute for the sigterm to finish processing.
  2. We attempt to sigterm not in the middle of a `Reconfigure` call. In this case, sigterm is processed and the server stops more or less immediately.

For a while there was different behavior depending on whether we sigtermed during the first `Reconfigure` call or not, but changes to how we update the resource graph has fixed that.

File: module/modmanager/manager.go
Comment (stuqdog) [reply module/modmanager/manager.go:81]: ahh good call! got too greedy with `s/inRecovery/inStartup` :)

**Comment by stuqdog** [comment]
> might be hard, but can we test that on manager closing, add/on unexpected exit handlers are properly exiting?

sure thing, will put something together and ping you when it's ready!

File: module/modmanager/manager.go
Comment (cheukt) [reply module/modmanager/manager.go:147]: cool, just slightly confused about the var name but it's been changed

File: module/modmanager/manager_test.go
Comment (benjirewis) [inline module/modmanager/manager_test.go:606]:
```suggestion
func TestGracefulShutdownWithMalformedModule(t *testing.T) {
	// This test ensures that module manager's `Add` can be interrupted by a `Close`
	// call correctly, and no OUE restart goroutines will continue beyond their `inStartup`
	// check. With our current design, `local_robot.Reconfigure` blocks the main thread,
	// so the manager will not be `Closed` while a module is being `Add`ed. Future work
	// (RSDK-4854) may change that though.
```

**Review by benjirewis — APPROVED** [review]
LGTM! Just a request for a comment in the test clarifying that it's not exactly a real world scenario.

**Comment by benjirewis** [comment]
I filed RSDK-4854 to track potential work for making `Reconfigure` non-blocking.

File: module/modmanager/manager_test.go
Comment (cheukt) [inline module/modmanager/manager_test.go:645]:
```suggestion
	// check that the OUE handler hasn't been called at this point
```
the current comment confused me a bit

**Review by cheukt — APPROVED** [review]
LGTM, thanks for looking into this!

**Pull request merged**
